File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const stack = [];
    
    for (let char of s) {
        if (['(', '[', '{'].includes(char)){
            stack.push(char);
        }
        if (char === ')') {
            if (stack[stack.length - 1] === '(') stack.pop();
            else return false;
        }
        if (char === ']') {
            if (stack[stack.length - 1] === '[') stack.pop();
            else return false;
        }
        if (char === '}') {
            if (stack[stack.length - 1] === '{') stack.pop();
            else return false;
        }
    }
    
    if (stack.length === 0) return true;
    else return false;
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        opened = set(['(', '{', '['])
        closed = set([')', '}', ']'])
        parens_map = {
            ')': '(',
            '}': '{',
            ']': '[',
        }

        for c in s:
            if c in opened:
                stack.append(c)
            if c in closed:
                if len(stack) == 0 or stack[-1] != parens_map[c]:
                    return False
                stack.pop()

        return len(stack) == 0